File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import (
	"errors"
)

// CollatzConjecture calculates how many steps are
// required to solve the Collatz Conjecture problem
func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
		return 0, errors.New("invalid input")
	}

	steps := 0

	for n > 1 {
		steps++
		if n%2 == 0 {
			n /= 2
		} else {
			n = 3*n + 1
		}
	}

	return steps, nil
}

File: c/collatz-conjecture/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int n) {
  if (n <= 0) return ERROR_VALUE;

  int i = 0;
  while (n != 1) {
    n = n & 1 ? (3 * n) + 1 : n >> 1;
    i++;
  }

  return i;
}
